### PR TITLE
IA-2245 Add DeletedNodepoolChecker

### PR DESCRIPTION
**Pull request by Qi77Qi**
https://broadworkbench.atlassian.net/browse/IA-2245

Scans through all active nodepools, and update DB accordingly when necessary
1. if nodepool doesn't exist in google anymore, mark it as DELETED and report this nodepool
2. If nodepool exist in google in ERROR, mark it as ERROR and report this nodepool
3. if nodepool exist in non ERROR state, do nothing and don't report the nodepool

### CODE REVIEW

File: zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DeletedNodepoolChecker.scala
Comment (kyuksel) [inline zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DeletedNodepoolChecker.scala:16]: Should the name reflect the fact that errored nodepools are being checked as well?

File: zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DeletedNodepoolChecker.scala
Comment (kyuksel) [inline zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DeletedNodepoolChecker.scala:44]: Shouldn't we mark apps as errored the way we do with deleted? If so, I wonder if it makes sense to have a `markNodepoolAndAppStatus()` method that takes an additional `status` parameter so it can be used for both cases.

File: zombie-monitor/src/test/scala/com/broadinstitute/dsp/zombieMonitor/DbReaderSpec.scala
Comment (kyuksel) [inline zombie-monitor/src/test/scala/com/broadinstitute/dsp/zombieMonitor/DbReaderSpec.scala:35]: typo
```suggestion
  it should "build activeNodepoolsQuery properly" in {
```

**Review by kyuksel — COMMENTED** [review]
Looks good overall. Added some comments/questions inline.

File: zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DeletedNodepoolChecker.scala
Comment (Qi77Qi) [reply zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DeletedNodepoolChecker.scala:44]: yep...realized that this morning :)

File: zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DbReader.scala
Comment (Qi77Qi) [reply zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DbReader.scala:37]: so we don't need custom mapping for case classes specifically, as long as there're `Get` defined for each individual fields (in [here](https://github.com/broadinstitute/leonardo-cron-jobs/blob/master/core/src/main/scala/com/broadinstitute/dsp/DbReaderImplicits.scala#L9))

**Comment by Qi77Qi** [comment]
@kyuksel thanks for reviewing..think I addressed all comments, let me know if you have more comments

File: zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DbReader.scala
Comment (kyuksel) [reply zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DbReader.scala:37]: Oh cool - thanks

File: zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DbReader.scala
Comment (kyuksel) [inline zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DbReader.scala:87]: ^^ looks like copy pasta 😛

File: zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DbReader.scala
Comment (kyuksel) [inline zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DbReader.scala:84]: Is there a reason why we couldn't have a single method 
```scala
updateNodepoolAndAppStatus(nodepoolId: Long, status: Status): F[Unit]
```
that we can use for updating the DB statuses for both `Deleted` and `Error` cases? Nit: `update` reads better to me than `mark`.

Also, any reason why not to have a `Status` ADT instead of using `String`?

File: zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DeletedOrErroredNodepoolChecker.scala
Comment (kyuksel) [inline zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DeletedOrErroredNodepoolChecker.scala:15]: I wonder if we should report the app as well 🤔

**Review by kyuksel — CHANGES_REQUESTED** [review]

File: zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DbReader.scala
Comment (Qi77Qi) [reply zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DbReader.scala:84]: nope..I think this was just because previous I wasn't updating them at the same time. will update

File: zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DbReader.scala
Comment (Qi77Qi) [reply zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DbReader.scala:87]: oh man, I was thinking the function is so simple and doesn't need a separate unit test for itself...apparently I was wrong

File: zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DeletedOrErroredNodepoolChecker.scala
Comment (Qi77Qi) [reply zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DeletedOrErroredNodepoolChecker.scala:15]: it'll be a bit more complicated if we want to report `App` as well since for now, for a given report file, we assume they're all the same type (like runtimes, disks, nodepools). so it'll require a bit more thinking how we organize things. so I think we should punt this for now.

Comment (kyuksel) [reply zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DeletedOrErroredNodepoolChecker.scala:15]: Punting for now makes sense. Should we create a ticket so we don't forget?

File: .github/scripts/createCommit.sh
Comment (kyuksel) [inline .github/scripts/createCommit.sh:38]: Nit: add new line

File: .github/scripts/createCommit.sh
Comment (kyuksel) [inline .github/scripts/createCommit.sh:1]: Is this script under `.github` directory because that's what's required for GitHub Actions?

File: .github/scripts/createCommit.sh
Comment (kyuksel) [inline .github/scripts/createCommit.sh:9]: Helpful to have the reference link above. Would be even better to comment on whether the help text on the link applies to this script and if not, clarify the distinctions here as the details of what's going on don't look straightforward. May also be good to add a comment as to how this script is used.

File: .github/workflows/createPullRequest.yml
Comment (kyuksel) [inline .github/workflows/createPullRequest.yml:35]: Looks like a good walkthrough topic candidate 😉

File: .github/workflows/unitTests.yml
Comment (kyuksel) [inline .github/workflows/unitTests.yml:25]: Nit: add new line

File: zombie-monitor/src/test/scala/com/broadinstitute/dsp/zombieMonitor/DbReaderSpec.scala
Comment (kyuksel) [inline zombie-monitor/src/test/scala/com/broadinstitute/dsp/zombieMonitor/DbReaderSpec.scala:148]: Nit
```suggestion
  it should "update nodepool and app status properly" in {
```

File: zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DbReader.scala
Comment (kyuksel) [inline zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DbReader.scala:75]: Should we also use a `Status` ADT the way we do in other codebases for more type safety?

**Review by kyuksel — COMMENTED** [review]
Thanks for addressing the previous comments. A few more inline 😊

File: zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DeletedOrErroredNodepoolChecker.scala
Comment (Qi77Qi) [reply zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DeletedOrErroredNodepoolChecker.scala:15]: on a second thought, I'm not sure if it's that useful to report `App`, if we have a specific app that we're wondering if it being updated properly, we can always go to its nodepool, and check the report to see if it was indeed updated by `janitor` (currently resource-validator).. all reports contain the source object, which is pretty easy to understand. If in the future, we have a use case where we do care about `App`, then I think it makes more sense to have a ticket...

File: .github/scripts/createCommit.sh
Comment (Qi77Qi) [reply .github/scripts/createCommit.sh:1]: not really...I'm just following examples on docs...don't worry about changes in `.github` yet...Things don't work properly just yet. I have a separate ticket working on this. I jsut figured I'll use this as a test. But I'm going to disable it before I merge this PR

File: .github/workflows/unitTests.yml
Comment (Qi77Qi) [reply .github/workflows/unitTests.yml:25]: what's the particular reason for adding new line btw? I never asked before, but figured I'd ask since you seem to suggest newline all the time

File: zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DbReader.scala
Comment (Qi77Qi) [reply zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DbReader.scala:75]: I feel ok without an ADT for now. If we publish leo models, they won't be needed here.

File: zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DeletedOrErroredNodepoolChecker.scala
Comment (kyuksel) [reply zombie-monitor/src/main/scala/com/broadinstitute/dsp/zombieMonitor/DeletedOrErroredNodepoolChecker.scala:15]: I thought it could debugging a bit faster in the case of UL questions, etc. but yea, it doesn't seem like a big deal right now since we have 1-to-1 mapping between apps and nodepools.

File: .github/scripts/createCommit.sh
Comment (kyuksel) [reply .github/scripts/createCommit.sh:1]: I see, okay. Might be good to do it as a separate PR next time 😊

File: .github/workflows/unitTests.yml
Comment (kyuksel) [reply .github/workflows/unitTests.yml:25]: Good question. In addition to the reasons [here](https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file), I also personally find the red marking GitHub adds to diffs distracting  🤷‍♂️

**Review by kyuksel — APPROVED** [review]

File: .github/scripts/createCommit.sh
Comment (Qi77Qi) [reply .github/scripts/createCommit.sh:1]: yea..sorry about this...oh actually, what I was thinking is if I fix the CI, then I don't have to manually fix the PR to `terra-helm`...that'll also prove CI works properly..so think I'm gonna try to get CI working properly before I merge this one (unless it takes too long to make CI work)

File: .github/workflows/createPullRequest.yml
Comment (Qi77Qi) [reply .github/workflows/createPullRequest.yml:35]: there's an item on walkthrough page about CI for this repo...so we can talk about this then

**Comment by Qi77Qi** [comment]
I realize I need to look into something for aou and won't be able to get to CI work immediately...so I'm going to merge this as is and come back to CI later

**Pull request merged**
